stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,82 @@
+#! input:  [v, ...]
+#! output: [v >> 48, ...]
+proc.shr_48
+    u32split
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 32, ...]
+proc.shr_32
+    u32split
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 16, ...]
+proc.shr_16
+    u32split
+    push.65536
+    mul
+    u32split
+    push.4294967296
+    mul
+    swap.2
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    add
+    add
+end
+
+#! Fetches the keyed value of a Sparse Merkle tree with the provided root.
+#! cycles: 91
+#! input:  [K, R, ...]
+#! output: [d, i, V, K, R, ...]
+export.get
+    # invoke the advice injector and fetch depth+index pair
+    # (18 cycles)
+    # [d, i, R, d, K, i, d, ...]
+    adv.smtget swapw adv_push.2 dup movdn.6 dup.1 movdn.11 dup movdn.12
+    # verify the opening to the root
+    # (9 cycles)
+    # [N, R, d, K, i, d, ...]
+    mtree_get
+    # prepare the depth to be used as domain for permutation
+    # (7 cycles)
+    # [0, 0, d, 0, N, R, K, i, d, ...]
+    push.0 movup.9 push.0.0
+    # prepare the key for permutation
+    # (4 cycles)
+    # [K, 0, 0, d, 0, N, R, K, i, d, ...]
+    dupw.3
+    # prepare the value for permutation
+    # (13 cycles)
+    # [V, K, 0, 0, d, 0, V, N, R, K, i, d, ...]
+    padw adv_loadw dupw swapw.3 movdnw.2
```

Comment:
What would happen here if the node is a root of an empty subtree? It wouldn't have a pre-image consisting of `V` and `K`, right?

Reply:
Yes, we can keep the discussion centralized here
https://github.com/0xPolygonMiden/miden-vm/pull/808#discussion_r1154010287